oonidata/oonidata/s3feeder.py

```diff
@@ -0,0 +1,508 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+"""
+Feeds reports from cans on public S3 bucke or local disk
+
+Explore bucket from CLI:
+AWS_PROFILE=ooni-data aws s3 ls s3://ooni-data/canned/2019-07-16/
+
+"""
+
+from datetime import date, timedelta, datetime
+from typing import Generator, Set, NamedTuple, Any, List
+from collections import namedtuple
+from pathlib import Path
+import itertools
+import logging
+import os
+import time
+import gzip
+import tarfile
+from multiprocessing import Pool
+
+import lz4.frame as lz4frame  # debdeps: python3-lz4
+import ujson
+
+# lz4frame appears faster than executing lz4cat: 2.4s vs 3.9s on a test file
+
+import boto3  # debdeps: python3-boto3
+from botocore import UNSIGNED as botoSigUNSIGNED
+from botocore.config import Config as botoConfig
+
+from .metrics import setup_metrics
+from .mytypes import MsmtTup  # msmt bytes, msmt dict, uid
+from .normalize import iter_yaml_msmt_normalized
+from .utils import trivial_id
+
+CAN_BUCKET_NAME = "ooni-data"
+MC_BUCKET_NAME = "ooni-data-eu-fra"
+MAX_PROCESS_COUNT = 24
+
+log = logging.getLogger("fastpath")
+metrics = setup_metrics(name="fastpath.s3feeder")
+
+
+def create_s3_client():
+    return boto3.client("s3", config=botoConfig(signature_version=botoSigUNSIGNED))
+
+s3 = create_s3_client()
+
+# suppress debug logs
+for x in ("urllib3", "botocore", "s3transfer"):
+    logging.getLogger(x).setLevel(logging.INFO)
+
+
+def load_multiple(fn: str) -> Generator[MsmtTup, None, None]:
+    """Load contents of legacy cans and minicans.
+    Decompress tar archives if found.
+    Yields measurements one by one as:
+        (string of JSON, None, uid) or (None, msmt dict, uid)
+    The uid is either taken from the filename or generated by trivial_id for
+    legacy cans
+    """
+    # TODO: split this and handle legacy cans and post/minicans independently
+    if fn.endswith(".tar.lz4"):
```

Comment:
These are not functions that are meant to be used by end users. We should actually never be giving end users the legacy formats, but just the published JSONLs. That way there is no need for them to ever call code dealing with the various old formats.

We do, however, need these in our codebase because we need to handle them as part of our data pipeline.